uni/lib/json.icn

```diff
@@ -991,3 +998,233 @@ end
+#
+procedure jsf_validate(jfname, sfname, strict, error)
+   local json_date, json_schema
+
+   jerror := ErrorHandler(error,strict)
+
+   json_data   := jtou(jfname , "f", strict, error)
+   json_schema := jtou(sfname , "f", strict, error)
+
+   return _jsvalidation(json_data, json_schema, json_schema)
+end
+
+procedure jss_validate(data, schema, strict, error)
+   local json_date, json_schema
+
+   jerror := ErrorHandler(error,strict)
+
+   json_data   := jtou(data , "s", strict, error)
+   json_schema := jtou(schema , "s", strict, error)
+
+   return _jsvalidation(json_data, json_schema, json_schema)
+end
+
+#<p>
+# 
+# Vaildate a specific value only at the curent level given a schema of that value.
+#
+# <[param x is the table]>
+# <[param val is the name of the value of a property]>
+# 
+# <[returns "" if the valid otherwise returs the error string]>
+#
+#</p>
+
+procedure validate_property(x, val)
+   z := ""
+   case x["type"] of {
+      "object" :{
+         if type(val) ~== "table" then {
+            return "expected object type, got: " || val
+            }
+         req := x["required"]
+         if \req then {
+            i := 1
+            cr := 0
+            while i <= *req do {
+               c := 0
+               every prop := key(val) do{
+                  if prop ~== req[i] then c := c + 1
+                  if prop == req[i] then cr := cr + 1
+                  }
+               if cr = *req then break
+                  if c =  *val then  z := "required properies missing"
+               i := i + 1
+               }
+            return z
+            }
+         maxProp := x["maxProperties"]
+         if \maxProp then write(space, "    maxProperties: ", left(maxProp, 20))
+         minProp := x["minProperties"]
+         if \minProp then write(space, "   minProperties: ", left(minProp, 20))
+         return z
+         }
+      "array": {
+         if type(val) ~== "list" then {
+            return "expected array type, got: " || image(val)
+            }
+         y := x["items"]
+         addItems := x["additionalItems"]
+         if \addItems then {
+            write(space, "   additionalItems: ")
+            dumpschema(x[additems])
+            }
+         minItems := x["minItems"]
+         if \minItems then write(space, "   minItems: ", left(minItems, 20))
+         uniqueItems := x["uniqueItems"]
+         if \uniqueItems then write(space, "   uniqueItems: ", left(uniqueItems, 20))
+         }
+      "string":{
+         if type(val) ~== "string" then {
+            return "expected string type, got: " || val
+            }
+         else{
+            enum := x["enum"]
+            if \enum then{
+               i := 0
+               c := 0
+               while i < *enum do {
+                  if val ~== enum[i] then c := c + 1
+                  i := i + 1
+                  }
+               if c ~= 0 then {
+                  return val || " is out of range, allowed values are: " || fullimage(enum)
```

Comment:
`fullimage()` has not changed since its introduction to the repo in 2001.  Should we be using `ximage` instead?
This comment applies to several places in this procedure.

Reply:
old doesn't mean less better! ;). `ximage()` is good for debugging, whereas `fullimage()` is good for user output. Example:

ximage:
```
L199 := list(3)
   L199[1] := "Alice"
   L199[2] := "Rayya"
   L199[3] := "Rayya"
```
fullimage:
```
["Alice","Rayya","Rayya"]
```

Reply:
Fair enough. In that case perhaps we should bring `fullimage` up to date: there are several things (e.g. patterns) that it doesn't deal with.

Reply:
add it to out growing TODO list! :)